enhancements/declarative-index-config.md

```diff
@@ -0,0 +1,574 @@
+
+Furthermore, a lot of the rules about the structure of an index are implicitly documented in various sources that are scattered. A best effort is being made to keep the rules updated in the form of a library in the [api repo](https://github.com/operator-framework/api/tree/master/pkg/validation) that is imported by tools like `opm`, `operator-courier` etc that perform index validation. However, like all tools that aspire to provide data validation, these tools, or api libraries that these tools import the rules from, have to be constantly maintained to keep up with the latest rule changes, while also maintaining older rules in previous releases. This introduces a lot of overhead in ensuring that olm is easy to use by users, while also leaving windows for lapses due to mis-coordination between different tools across different versions.
+
+Since indexes are manipulated by making additive changes to it, the way an index can be reproduced today is my managing a knowledge base of the series of changes made to the index (i.e storing the `opm` commands chronologically) along with a set of intermediary artifacts (images of the index with their corresponding digest etc). This introduces a lot of overhead in maintaining indexes, and as a by product has significantly burdened the `opm` tool with feature requests to add an increasing number of sub-commands to allow index authors to manipulate their indexes in various ways. 
+
+### Goals
+
+A human consumable representation of packages in an index in the form of json/yaml will: 
+
+1. Allow for editing of package level information without rebuilding and republishing artifacts(remove bundle/s from a channel, switch bundle/s to a different channel, edit the upgrade graph in a channel etc).
+2. Allow algebraic operations to be performed on indexes in a seamless way(combine two indexes, split an index, remove bundles from the index, etc).
```

Comment:
combining and splitting indexes don't seem like things we explicitly need, is there a use case that ties to those?  they seem to add exceptional complexity to the tooling that must ultimately manage this config (what do you do when splitting an index results in broken upgrade graphs?)

Reply:
It's somewhat implicit in a lot of the discussions around indexes and pipelines. If we want owners of a particular operator to be able to own and modify their own update graph, we need to be able to take just that component of an index and incorporate it into a larger index.

If you think of that bit of the graph as "an index with only one operator in it" the requirement for combining indexes makes some sense.

Reply:
Yeah, I have to +1 Evan here. The core requirement is "I just want to edit my package's graph without having to worry about all the other unrelated packages in the index"

Reply:
@bparees  That is a really good question, although, if we're to facilitate update of package information using the config, we can get into situations where the update graph is broken even without the use case of combining or splitting indexes(eg manually removing a bundle from `bundles` when it is part of an upgrade path). I don't think we can avoid taking on some additional work for the tooling (eg visualizing an upgrade graph for a package, verifying the upgrade graph is not broken( due to missing bundles etc) before accepting the updated config in the index) if we want to facilitate updating the index structure using the config. (fyi [this enhancement](https://github.com/operator-framework/enhancements/pull/26) is slated to be the place for introducing all the changes needed in the tooling to assist the config)

Reply:
are two packages ever related/connected in any way in a catalog today?

assuming not, i would reframe it as "extract/remove/edit/replace" the entire upgrade graph for a single package within a given catalog".  That is a much simpler concept than "algebraic operations on catalogs"